### PR TITLE
Feature/encode decode dynamic data oop

### PR DESCRIPTION
**Pull request by etienneroudeix**

### CODE REVIEW

File: src/EthereumRawTx/Abi/Abi.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        return $this->constructor;
+    }
+
+    public function getFunctionByPrototype($prototype): FunctionItem
+    {
+        $hash = substr($hash, 0, 8);
+
+        return $this->functions[$hash] ?? null;
+    }
```
Comment (NeverEnding8) [inline src/EthereumRawTx/Abi/Abi.php]: In getFunctionByPrototype function,  the variable $hash is not correct name : $prototype

Duplacate function with getFunctionByPrototypeHash ? i don't see any call to getFunctionByPrototype()

File: src/EthereumRawTx/Abi/ParamType.php [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+                if ($value instanceof Buffer === false) {
+                    throw new \Exception("Value must be an Buffer");
+                }
+                return str_pad($value->getHex(), 64, '0', STR_PAD_LEFT);
+
+            case 'string':
+            case 'bytes':
+                if($this->isDynamic()) {
+                    if ($value instanceof Buffer === true) {
+                        throw new \Exception('TODO');
```
Comment (NeverEnding8) [inline src/EthereumRawTx/Abi/ParamType.php]: Why Todo ? i would like to send Hexadecimal bytes whithout string encoding

File: src/EthereumRawTx/Abi/ParamType.php [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                        // TODO
+                        //return $this->encodeParam('uint', Buffer::int(strlen($value->getHex()) / 2)).$value->getHex();
+                    }
+                    return StringEncoder::encode($value);
+                }
+
+                if ($value instanceof Buffer === true) {
+                    throw new \Exception('TODO');
```
Comment (NeverEnding8) [inline src/EthereumRawTx/Abi/ParamType.php]: Why Todo ? i would like to send Hexadecimal bytes whithout string encoding

File: src/EthereumRawTx/Encoder/StringEncoder.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * @param string $string
+     * @param int $lenghtBytes
+     * @return string
+     * @throws \Exception
+     */
+    public static function encode(string $string, int $lenghtBytes = 0)
+    {
+        $return = '';
+        // todo $lenghtBytes > 0
```
Comment (NeverEnding8) [inline src/EthereumRawTx/Encoder/StringEncoder.php]: todo ? task already done ?

**Review by NeverEnding8 — CHANGES_REQUESTED** [review]

File: src/EthereumRawTx/Abi/Abi.php (hunk 1 above)
Comment (etienneroudeix) [reply src/EthereumRawTx/Abi/Abi.php]: Cette fonction n'a pas de raison d'exister, je la supprime

File: src/EthereumRawTx/Encoder/StringEncoder.php (hunk 4 above)
Comment (etienneroudeix) [reply src/EthereumRawTx/Encoder/StringEncoder.php]: indeed.

File: src/EthereumRawTx/Abi/ParamType.php (hunk 2 above)
Comment (etienneroudeix) [reply src/EthereumRawTx/Abi/ParamType.php]: todo because i did not write specs. Done now.

File: src/EthereumRawTx/Abi/ParamType.php (hunk 3 above)
Comment (etienneroudeix) [reply src/EthereumRawTx/Abi/ParamType.php]: todo because i did not write specs. Done now.

**Comment by etienneroudeix** [comment]
@NeverEndingLoopMan Patched.

**Review by NeverEnding8 — APPROVED** [review]

**Pull request merged**